opt: lock all column families for new SELECT FOR UPDATE

michae2 opened the pull request:
**opt: lock all column families for new SELECT FOR UPDATE**

The new implementation of SELECT FOR UPDATE used for Read Committed isolation starts as a lock operator built from a table scan, which becomes a lookup join during execbuild. This lock operator was re-using the same table and column references as the original table scan, which caused the resulting lookup join to only lock the first column family of the table.

Instead, we need to use fresh table and column references for the lock operator, so that the resulting lookup join is able to control which column families it locks.

For now we choose to lock all column families, even if some column family would not normally be read by the select statement. We do this to more closely match Postgres behavior for statements like:

```
SELECT 1 FROM foo FOR UPDATE
```

Fixes: https://github.com/cockroachdb/cockroach/issues/115014

Release note (sql change): The new SELECT FOR UPDATE implementation used under Read Committed isolation (and under Serializable isolation when optimizer_use_lock_op_for_serializable is true) now locks all column families instead of only the first column family.

---

**sql: do not skip key col families when performing a locked read**

When determining which column families to read from an index for a set of requested columns, the span splitter has an optimization: it skips over the families for requested columns that are completely encoded in the index key, because these columns can be decoded from the key of *any* column family. For example, for the following query we do not need to read column families foo and bar:

```
CREATE TABLE abc (
  a INT NOT NULL,
  b INT NOT NULL,
  c INT NOT NULL,
  PRIMARY KEY (a, b),
  FAMILY foo (a),
  FAMILY bar (b),
  FAMILY baz (c)
);
SELECT * FROM abc WHERE a = 5 AND b = 6;
```

When the read has a side-effect, however (e.g. SELECT FOR UPDATE) we need to lock all of the requested column families. To do so, this commit adds a variant of `span.MakeSplitter` for locked reads, called `span.MakeSplitterForSideEffect`, which does not use this optimization.

Fixes: #115014

Release note (sql change): Fix a bug where SELECT FOR UPDATE under Read Committed isolation on multi-column-family tables was not locking column families containing only key columns.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/116170)
<!-- Reviewable:end -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice fix!

> Reviewed 1 of 1 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-:-NlUozQgCY8MioBqqdjX:b-46lt7k)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1150 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NlUCAAk2IG2TXbEr1uo:-NlUCAAk2IG2TXbEr1up:bh55veh) ([raw file](https://github.com/cockroachdb/cockroach/blob/e80aca6c161a1298d0cd0a5ad36baa68f3e126e5/pkg/sql/opt/exec/execbuilder/mutation.go#L1150)):*
> ```Go
> 
> 	// In some simple cases we can push the locking down into the input operation
> 	// instead of adding what would be a redundant lookup join.
> ```

Does this optimization need to change to ensure that the input is scanning the same columns as are required by the lock?
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NlU7uvs9O1bR3ZeGkT9:-NlU7uvt1UpQnm1pKEhG:b-wuxf63) ([raw file](https://github.com/cockroachdb/cockroach/blob/e80aca6c161a1298d0cd0a5ad36baa68f3e126e5/pkg/sql/opt/exec/execbuilder/mutation.go#L1160)):*
> ```Go
> 	switch input := lock.Input.(type) {
> 	case *memo.ScanExpr:
> 		if md.Table(input.Table) == md.Table(lock.Table) && input.Index == cat.PrimaryIndex {
> ```

[nit] I think this check is correct, but occasionally I've seen the catalog tables be different pointers even for the same table. I think it would be best to use `md.Table(input.Table).ID() == md.Table(lock.Table).ID()`
___
*[`pkg/sql/opt/optbuilder/locking.go` line 398 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NlU9QQa8ceZf6yf1Ahp:-NlU9QQb-Qot7UgT1Ofo:b-nhjji0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e80aca6c161a1298d0cd0a5ad36baa68f3e126e5/pkg/sql/opt/optbuilder/locking.go#L398)):*
> ```Go
> 		oldColID := lb.table.ColumnID(i)
> 		newColID := newTabID.ColumnID(i)
> 		colMap.Set(int(oldColID), int(newColID))
> ```

[nit] ColMaps can have poor performance for queries with a lot of columns. Can we follow the example of `DuplicateColumnIDs` instead? Or better yet, call into a shared method on the factory that returns the new table and column IDs?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116170)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed` line 403 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NlUpuls5qVazlgpPowI:-NlUpuls5qVazlgpPowJ:b-79s1gu) ([raw file](https://github.com/cockroachdb/cockroach/blob/e80aca6c161a1298d0cd0a5ad36baa68f3e126e5/pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed#L403)):*
> ```
> 
> statement ok
> CREATE TABLE abcde (a INT NOT NULL PRIMARY KEY, b INT NOT NULL, c INT NOT NULL, d INT NOT NULL, e INT NOT NULL, FAMILY (a, b), FAMILY (c), FAMILY (d), FAMILY (e))
> ```

A case with a nullable column might be interesting. I think attempting to lock (only) families with id > 0 and all nullable columns will also lock family 0.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116170)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1150 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NlUCAAk2IG2TXbEr1uo:-NlUqqwH0orRV2v9l3p9:b-pxpw7k) ([raw file](https://github.com/cockroachdb/cockroach/blob/e80aca6c161a1298d0cd0a5ad36baa68f3e126e5/pkg/sql/opt/exec/execbuilder/mutation.go#L1150)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Does this optimization need to change to ensure that the input is scanning the same columns as are required by the lock?
</blockquote></details>

Or at least, should we check that the input is scanning _at least_ the columns that need to be locked?


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NlU7uvs9O1bR3ZeGkT9:-NldVQLM-VmA2dIW_NVU:b9n7oeh) ([raw file](https://github.com/cockroachdb/cockroach/blob/e80aca6c161a1298d0cd0a5ad36baa68f3e126e5/pkg/sql/opt/exec/execbuilder/mutation.go#L1160)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] I think this check is correct, but occasionally I've seen the catalog tables be different pointers even for the same table. I think it would be best to use `md.Table(input.Table).ID() == md.Table(lock.Table).ID()`
</blockquote></details>

`cat.Table.ID()` returns the stable ID of the underlying table in the schema, not the logic table in the plan. So in a query that joins a table to itself, could that cause problems here?


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NlU7uvs9O1bR3ZeGkT9:-NldWCj5CW-DST2ptN0Z:b-42ejti) ([raw file](https://github.com/cockroachdb/cockroach/blob/e80aca6c161a1298d0cd0a5ad36baa68f3e126e5/pkg/sql/opt/exec/execbuilder/mutation.go#L1160)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

`cat.Table.ID()` returns the stable ID of the underlying table in the schema, not the logic table in the plan. So in a query that joins a table to itself, could that cause problems here?
</blockquote></details>

I suppose that's what we want here, since this code was comparing the `opt.TableID`s before?


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116170)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball, @michae2, and @nvanbenschoten)

___
*[`pkg/sql/opt/optbuilder/locking.go` line 409 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NldX_IC3HrBhliXoK0J:-NldX_IC3HrBhliXoK0K:bh5xi94) ([raw file](https://github.com/cockroachdb/cockroach/blob/e80aca6c161a1298d0cd0a5ad36baa68f3e126e5/pkg/sql/opt/optbuilder/locking.go#L409)):*
> ```Go
> 		oldColID := inScope.cols[i].id
> 		if newColID, ok := colMap.Get(int(oldColID)); ok {
> 			cols.Add(opt.ColumnID(newColID))
> ```

Is `cols` supposed to include columns from both `inScope` and the new table?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
TFTRs!

Drew's questions brought up a number of problems that I have been working on. Hold off on reviewing for a bit, will push today with fixes.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-:-NldfChoEE1Nu8YMmNuA:b-j0ll49)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @nvanbenschoten)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1150 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NlUCAAk2IG2TXbEr1uo:-NldeuDS2k4eiqx1WSiG:bgoopih) ([raw file](https://github.com/cockroachdb/cockroach/blob/e80aca6c161a1298d0cd0a5ad36baa68f3e126e5/pkg/sql/opt/exec/execbuilder/mutation.go#L1150)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Or at least, should we check that the input is scanning _at least_ the columns that need to be locked?
</blockquote></details>

Yes, great point, this optimization does now need to check columns.
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NlU7uvs9O1bR3ZeGkT9:-Nldf-PVCZ_2cuU1w23j:b2b0ggd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e80aca6c161a1298d0cd0a5ad36baa68f3e126e5/pkg/sql/opt/exec/execbuilder/mutation.go#L1160)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I suppose that's what we want here, since this code was comparing the `opt.TableID`s before?
</blockquote></details>

Yes, exactly, a query that joins a table to itself now has a problem, which I am currently fixing.


<!-- Sent from Reviewable.io -->

michae2 commented:
I have realized there's a problem with this approach: statements like `SELECT 1 FROM foo FOR UPDATE` won't lock _any_ column families. For 23.2 I am going to switch to locking _all_ column families, which will simplify this a lot, and maybe we can revisit this in 24.1.

michae2 reviewed (commented):
Actually, I changed my mind, because locking individual column families seems really useful. I plan to make `SELECT 1 FROM foo FOR UPDATE` lock the first column family, as it does now.

I think this new version will work, but hold off on reviewing until I add more tests.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-:-NlgIPLl0RKvJXQQor0w:b7nar2t)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116170)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, and @nvanbenschoten)

___
*[`pkg/sql/opt/optbuilder/locking.go` line 405 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NllVqq29BPn7y8fHc9d:-NllVqq29BPn7y8fHc9e:bkkgosd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c025af6df66d0ae283b4a3d5e0f0d6bad23765f8/pkg/sql/opt/optbuilder/locking.go#L405)):*
> ```Go
> 	// empty. In this case we will only lock the first column family.
> 	var lockCols opt.ColSet
> 	for _, col := range inScope.cols {
> ```

Hmm, ran into another quandary. This method of picking the columns to lock from the projection scope means that a statement like:

```sql
SELECT a + b FROM foo FOR UPDATE;
```

won't actually lock the column families of a and b, but will instead only lock the first column family, just as I'm doing for `SELECT 1 FROM foo FOR UPDATE`. Maybe even scarier is a statement like:

```sql
SELECT a::int FROM foo FOR UPDATE;
```

which really looks like it should lock column family a, but does not.

I think this is too subtle for people to work with safely. I'm going to go back to locking all column families.


<!-- Sent from Reviewable.io -->

michae2 committed:
opt: lock all column families for new SELECT FOR UPDATE

The new implementation of SELECT FOR UPDATE used for Read Committed
isolation starts as a lock operator built from a table scan, which
becomes a lookup join during execbuild. This lock operator was re-using
the same table and column references as the original table scan, which
caused the resulting lookup join to only lock the first column family of
the table.

Instead, we need to use fresh table and column references for the lock
operator, so that the resulting lookup join is able to control which
column families it locks.

For now we choose to lock all column families, even if some column
family would not normally be read by the select statement. We do this to
more closely match Postgres behavior for statements like:

```
SELECT 1 FROM foo FOR UPDATE
```

Fixes: #115014

Release note (sql change): The new SELECT FOR UPDATE implementation used
under Read Committed isolation (and under Serializable isolation when
optimizer_use_lock_op_for_serializable is true) now locks all column
families instead of only the first column family.

michae2 reviewed (commented):
Changed it to lock all column families.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-:-Nlvn2g36EX0zKD75OID:b-ofpquj)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, and @nvanbenschoten)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1150 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NlUCAAk2IG2TXbEr1uo:-NlvfchZD5mUKyXvirka:bkdrjto) ([raw file](https://github.com/cockroachdb/cockroach/blob/e80aca6c161a1298d0cd0a5ad36baa68f3e126e5/pkg/sql/opt/exec/execbuilder/mutation.go#L1150)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Yes, great point, this optimization does now need to check columns.
</blockquote></details>

I made the optimization only apply to single-column-family tables. This saves us from having to compare the columns produced by the input to the columns locked, because there's only one column family in play.
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NlU7uvs9O1bR3ZeGkT9:-Nlvg2HJ8tyHENPMJmMK:bj0dwwm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e80aca6c161a1298d0cd0a5ad36baa68f3e126e5/pkg/sql/opt/exec/execbuilder/mutation.go#L1160)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Yes, exactly, a query that joins a table to itself now has a problem, which I am currently fixing.
</blockquote></details>

I switched back to comparing the exact `opt.TableID` (by saving it in the `KeySource` field of the `LockPrivate`).
___
*[`pkg/sql/opt/optbuilder/locking.go` line 398 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NlU9QQa8ceZf6yf1Ahp:-NlvhNqn0T6k_3FJEr13:b-7huy36) ([raw file](https://github.com/cockroachdb/cockroach/blob/e80aca6c161a1298d0cd0a5ad36baa68f3e126e5/pkg/sql/opt/optbuilder/locking.go#L398)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] ColMaps can have poor performance for queries with a lot of columns. Can we follow the example of `DuplicateColumnIDs` instead? Or better yet, call into a shared method on the factory that returns the new table and column IDs?
</blockquote></details>

I ended up putting _every_ column of the duplicated table into the set, instead of using the scope to pick, so the ColMap went away. (This is done to ensure we lock _all_ column families.)
___
*[`pkg/sql/opt/optbuilder/locking.go` line 409 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NldX_IC3HrBhliXoK0J:-Nlvmg_k38b06m7cIXp9:b-rpraz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e80aca6c161a1298d0cd0a5ad36baa68f3e126e5/pkg/sql/opt/optbuilder/locking.go#L409)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Is `cols` supposed to include columns from both `inScope` and the new table?
</blockquote></details>

I added the `LockCols` field to make this more clear: `Cols` is essentially passthrough columns from `inScope` and `LockCols` has columns from the new table.
___
*[`pkg/sql/opt/optbuilder/locking.go` line 405 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NllVqq29BPn7y8fHc9d:-NlviCDF3a0-7Fo1qX8D:b-a4bm71) ([raw file](https://github.com/cockroachdb/cockroach/blob/c025af6df66d0ae283b4a3d5e0f0d6bad23765f8/pkg/sql/opt/optbuilder/locking.go#L405)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Hmm, ran into another quandary. This method of picking the columns to lock from the projection scope means that a statement like:

```sql
SELECT a + b FROM foo FOR UPDATE;
```

won't actually lock the column families of a and b, but will instead only lock the first column family, just as I'm doing for `SELECT 1 FROM foo FOR UPDATE`. Maybe even scarier is a statement like:

```sql
SELECT a::int FROM foo FOR UPDATE;
```

which really looks like it should lock column family a, but does not.

I think this is too subtle for people to work with safely. I'm going to go back to locking all column families.
</blockquote></details>

I think we should introduce a syntax something like `SELECT ... FOR UPDATE OF foo (a, b)` to make it unambiguous which column families are being locked, instead of trying to rely on the select clause.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Great job figuring all this out!

I think there's one last problem - when the primary key is fully specified (e.g. not looking up with a prefix), and a column family is fully contained in the key column set, we can skip scanning that family and perform `get` requests on the others. Example:
```
CREATE TABLE abc (a INT NOT NULL, b INT NOT NULL, c INT NOT NULL, PRIMARY KEY (a, b), FAMILY foo (a), FAMILY bar (b), FAMILY baz (c));
EXPLAIN ANALYZE (DEBUG) SELECT * FROM abc WHERE a = 5 AND b = 6;
--From plan.txt:
• scan
  columns: (a int, b int, c int)
  nodes: n1
  actual row count: 0
  vectorized batch count: 0
  KV time: 1ms
  KV contention time: 0µs
  KV rows decoded: 0
  KV pairs read: 0
  KV bytes read: 0 B
  KV gRPC calls: 1
  estimated max memory allocated: 20 KiB
  sql cpu time: 35µs
  MVCC step count (ext/int): 0/0
  MVCC seek count (ext/int): 1/1
  estimated row count: 1 (missing stats)
  table: abc@abc_pkey
  spans: /5/6/2/1
```
Note from the spans that only the `baz` column family was scanned. The span-splitting logic is called from `DistSQLPlanner.createPlanForLookupJoin`. We could potentially make that a separate fix - whatever's easiest.

> Reviewed 1 of 6 files at r3, 4 of 6 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-:-NlvwjeFBmfgT6djjQnF:b-3hcm14)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @michae2, and @nvanbenschoten)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NlU7uvs9O1bR3ZeGkT9:-NlvrsMu1k8xLj5I8hxw:bxpvtpf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e80aca6c161a1298d0cd0a5ad36baa68f3e126e5/pkg/sql/opt/exec/execbuilder/mutation.go#L1160)):*
>I switched back to comparing the exact `opt.TableID` (by saving it in the `KeySource` field of the `LockPrivate`).

SGTM. Seems best to limit the scope of the optimization for now.
___
*[`pkg/sql/opt/optbuilder/locking.go` line 405 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NllVqq29BPn7y8fHc9d:-NlvreIyELjzKuslXq23:b-2ibxtc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c025af6df66d0ae283b4a3d5e0f0d6bad23765f8/pkg/sql/opt/optbuilder/locking.go#L405)):*
>I think we should introduce a syntax something like `SELECT ... FOR UPDATE OF foo (a, b)` to make it unambiguous which column families are being locked, instead of trying to rely on the select clause.

I like this idea.
___
*[`pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed` line 403 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NlUpuls5qVazlgpPowI:-NlvwGfeEGKfqgqjFEEr:b-dmrw9a) ([raw file](https://github.com/cockroachdb/cockroach/blob/e80aca6c161a1298d0cd0a5ad36baa68f3e126e5/pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed#L403)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

A case with a nullable column might be interesting. I think attempting to lock (only) families with id > 0 and all nullable columns will also lock family 0.
</blockquote></details>

A nullable case might still be interesting, but seems less important now.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo Drew's comment

> Reviewed 4 of 6 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-:-Nlxh2KO3ppBaIktYiI1:bdlxrty)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @DrewKimball, @michae2, and @nvanbenschoten)

___
*[`pkg/sql/opt/ops/mutation.opt` line 384 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NlxgFzeBU_plmY5Srdz:-NlxgFzeBU_plmY5Sre-:b-69ez6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/29380c91de14b600ef972938bb28df8125145e5b/pkg/sql/opt/ops/mutation.opt#L384)):*
> ```
> 
>     # KeySource identifies the source of the key columns.
>     KeySource TableID
> ```

In a follow-up, consider adding some assertions about the interesting invariants of this expression in check_expr.go, like:

* Assert that all columns in KeyCols are from the KeySource table.
* Assert that all columns in LockCols are from Table.
* Assert that Cols contains KeyCols.
* Assert that Cols does not intersect LockCols.
* Assert the order of KeyCols.
___
*[`pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed` line 400 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NlxhES57OUE3Q4R63wm:-NlxhES57OUE3Q4R63wn:b-nss3d4) ([raw file](https://github.com/cockroachdb/cockroach/blob/29380c91de14b600ef972938bb28df8125145e5b/pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed#L400)):*
> ```
> 
> statement ok
> CREATE TABLE abcde (a INT NOT NULL PRIMARY KEY, b INT NOT NULL, c INT NOT NULL, d INT NOT NULL, e INT NOT NULL, FAMILY (a, b), FAMILY (c), FAMILY (d), FAMILY (e))
> ```

nit: The table would be easier to read if split on multiple lines.


<!-- Sent from Reviewable.io -->

michae2 committed:
sql: do not skip key col families when performing a locked read

When determining which column families to read from an index for a set
of requested columns, the span splitter has an optimization: it skips
over the families for requested columns that are completely encoded in
the index key, because these columns can be decoded from the key of
*any* column family. For example, for the following query we do not need
to read column families foo and bar:

```
CREATE TABLE abc (
  a INT NOT NULL,
  b INT NOT NULL,
  c INT NOT NULL,
  PRIMARY KEY (a, b),
  FAMILY foo (a),
  FAMILY bar (b),
  FAMILY baz (c)
);
SELECT * FROM abc WHERE a = 5 AND b = 6;
```

When the read has a side-effect, however (e.g. SELECT FOR UPDATE) we
need to lock all of the requested column families. To do so, this commit
adds a variant of `span.MakeSplitter` for locked reads, called
`span.MakeSplitterForSideEffect`, which does not use this optimization.

Fixes: #115014

Release note (sql change): Fix a bug where SELECT FOR UPDATE under Read
Committed isolation on multi-column-family tables was not locking column
families containing only key columns.

michae2 reviewed (commented):
> I think there's one last problem - when the primary key is fully specified (e.g. not looking up with a prefix), and a column family is fully contained in the key column set, we can skip scanning that family and perform `get` requests on the others.

Wow, great catch!

I pushed a second commit fixing that behavior.

Interestingly, I was not able to come up with a scenario where this lack of a lock on family `bar` would cause a problem, because any change to `b` causes a delete and a put, which touch all column families. And we always have at least the first column family locked. But maybe someone else can come up with a scenario?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-:-Nlzhq8I-7CawQ8mM9Tm:byks37i)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @DrewKimball, @mgartner, and @nvanbenschoten)

___
*[`pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed` line 403 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NlUpuls5qVazlgpPowI:-NlziKYa2OEkuR7bNK_1:bek2f40) ([raw file](https://github.com/cockroachdb/cockroach/blob/e80aca6c161a1298d0cd0a5ad36baa68f3e126e5/pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed#L403)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

A nullable case might still be interesting, but seems less important now.
</blockquote></details>

Done. (column d)
___
*[`pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed` line 400 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NlxhES57OUE3Q4R63wm:-NlziNKYE1jtu_J5OY__:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/29380c91de14b600ef972938bb28df8125145e5b/pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed#L400)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: The table would be easier to read if split on multiple lines.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
My fix was a little too aggressive, and broke a test added by https://github.com/cockroachdb/cockroach/pull/42572 for FK checks on multi-column-family tables. Now I'm only modifying the behavior for lookup joins under read committed isolation.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-:-Nm01xMI3l8o1eCO4BUU:b-t5laaa)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @DrewKimball, @mgartner, and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Had some suggestions/comments, but I'm happy with this as-is if we want to merge.

>Interestingly, I was not able to come up with a scenario where this lack of a lock on family `bar` would cause a problem, because any change to `b` causes a delete and a put, which touch all column families. And we always have at least the first column family locked. But maybe someone else can come up with a scenario?

That's a good point. Since the problem case requires the column to be part of the primary key, any mutations to that column involve all families automatically. So, maybe the only problematic case is two locking reads.

> Reviewed 2 of 2 files at r6, 4 of 7 files at r7, 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-:-Nm0XUKe3vRnjFoEnGR5:buwk3fe)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @michae2, and @nvanbenschoten)

___
*[`pkg/sql/distsql_physical_planner.go` line 3120 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-Nm0_MhW0DwJrTACeVC9:-Nm0_MhW0DwJrTACeVCA:b79mt8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/40d497a7e8d97465b308b4e999a16f9ce715134b/pkg/sql/distsql_physical_planner.go#L3120)):*
> ```Go
> 	if joinReaderSpec.LockingStrength != descpb.ScanLockingStrength_FOR_NONE &&
> 		planCtx.ExtendedEvalCtx.TxnIsoLevel != isolation.Serializable {
> 		splitter = span.MakeSplitterForSideEffect(n.table.desc, n.table.index, fetchOrdinals)
> ```

I'm happy with the current implementation, so please feel free to ignore, but one thought: we could just check that this condition is false before calling `MakeSplitter`, since a nil `SplitFamilyIDs` indicates that no splitting will be performed and all column families should be scanned. Then, we could even just get rid of the entire first commit, since we would be guaranteed to scan all column families, no need for the `LockCols` field or `TableID` duplication.
___
*[`pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed` line 484 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-Nm0YrDQ8EMwX1YTjr5W:-Nm0YrDQ8EMwX1YTjr5X:bgmyu2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e78312c138b1c9d000fa471e88ec5387020765/pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed#L484)):*
> ```
> SELECT d FROM abcde WHERE a = 45 AND b = 46 FOR UPDATE
> ----
> NULL
> ```

BTW, one of the things that makes this case interesting is that AFAIK we don't actually write a KV entry for entirely NULL column families other than family 0. So, if and when we decide to do something other than locking all families, we'll want to make sure the persistent locks interact well with that.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116170)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, @michae2, and @nvanbenschoten)

___
*[`pkg/sql/opt/ops/mutation.opt` line 378 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-Nm1_WAGD1Z39s-yDdeC:-Nm1_WAGD1Z39s-yDdeD:b-uswjxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/40d497a7e8d97465b308b4e999a16f9ce715134b/pkg/sql/opt/ops/mutation.opt#L378)):*
> ```
> 
> [Private]
> define LockPrivate {
> ```

nit: This is all a bit confusing for someone without the context here. Can you add an example, maybe using the query above the `Lock` definition, and specify what `Table`, `KeySource`, `KeyCols` etc are in this case?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
TFTRs! (Looks like @mgartner has drafts so I will wait.)

Also had a [conversation with Nathan in slack](https://cockroachlabs.slack.com/archives/C051DMXEGMV/p1702999350959479) about locking all column families.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-:-Nm2eD9sC3AAPnOTU936:bh85sbm)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @DrewKimball, @mgartner, and @nvanbenschoten)

___
*[`pkg/sql/distsql_physical_planner.go` line 3120 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-Nm0_MhW0DwJrTACeVC9:-Nm2K9gY-8Dt-oshBYXV:b-mv75mr) ([raw file](https://github.com/cockroachdb/cockroach/blob/40d497a7e8d97465b308b4e999a16f9ce715134b/pkg/sql/distsql_physical_planner.go#L3120)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I'm happy with the current implementation, so please feel free to ignore, but one thought: we could just check that this condition is false before calling `MakeSplitter`, since a nil `SplitFamilyIDs` indicates that no splitting will be performed and all column families should be scanned. Then, we could even just get rid of the entire first commit, since we would be guaranteed to scan all column families, no need for the `LockCols` field or `TableID` duplication.
</blockquote></details>

Interesting. This is a really good point, and not something I thought of.

I'm a little scared to do this, because by the time we're doing physical planning we don't know if the locking lookup join came from a SELECT FOR UPDATE statement or a FK check. I believe FK checks are still only locking column family 0, and at least back in the time of #42572 we were trying to preserve this behavior. I think using a nil `SplitFamilyIDs` would cause all locking lookup joins, including those used for FK checks, to lock all column families.
___
*[`pkg/sql/opt/ops/mutation.opt` line 378 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-Nm1_WAGD1Z39s-yDdeC:-Nm2LPdF8or-H_h7EcVU:bdtem5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/40d497a7e8d97465b308b4e999a16f9ce715134b/pkg/sql/opt/ops/mutation.opt#L378)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: This is all a bit confusing for someone without the context here. Can you add an example, maybe using the query above the `Lock` definition, and specify what `Table`, `KeySource`, `KeyCols` etc are in this case?
</blockquote></details>

Done, lmk if that is helpful.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 6 files at r3, 2 of 2 files at r6, 4 of 7 files at r7, 3 of 3 files at r8, 7 of 7 files at r9, 6 of 6 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-:-Nm2uLjNEK4rBaKUT-8u:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @michae2, and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

michae2 commented:
Thanks!

bors r=DrewKimball,mgartner

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116170)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[`pkg/sql/opt/optbuilder/locking.go` line 405 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-NllVqq29BPn7y8fHc9d:-Nm3q-qK3Si7m0LpGFVB:b-dkrosq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c025af6df66d0ae283b4a3d5e0f0d6bad23765f8/pkg/sql/opt/optbuilder/locking.go#L405)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

>I think we should introduce a syntax something like `SELECT ... FOR UPDATE OF foo (a, b)` to make it unambiguous which column families are being locked, instead of trying to rely on the select clause.

I like this idea.
</blockquote></details>

Filed https://github.com/cockroachdb/cockroach/issues/116838 about this.
___
*[`pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed` line 484 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/116170#-Nm0YrDQ8EMwX1YTjr5W:-Nm3q8MLD9_zK0EIGZG7:b-j31qi2) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e78312c138b1c9d000fa471e88ec5387020765/pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed#L484)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

BTW, one of the things that makes this case interesting is that AFAIK we don't actually write a KV entry for entirely NULL column families other than family 0. So, if and when we decide to do something other than locking all families, we'll want to make sure the persistent locks interact well with that.
</blockquote></details>

Yes, and as it turns out this is still an issue. Filed https://github.com/cockroachdb/cockroach/issues/116836.


<!-- Sent from Reviewable.io -->
